src/translate.py

```diff
@@ -2608,6 +2627,137 @@ def format_f64_imm(num: int) -> str:
+    # sign/size information to stay simpler. The result will be made with BinaryOp.int()
+    # regardless of input types.
+    expr = original_expr
+    left_expr = early_unwrap_ints(expr.left)
+    right_expr = early_unwrap_ints(expr.right)
+    divisor_shift = 0
+
+    # Fold `/` with `>>`: ((x / N) >> M) --> x / (N << M)
+    if (
+        isinstance(left_expr, BinaryOp)
+        and left_expr.op == "/"
+        and isinstance(left_expr.right, Literal)
+        and expr.op == ">>"
+        and isinstance(right_expr, Literal)
+    ):
+        new_denom = left_expr.right.value << right_expr.value
+        if new_denom < (1 << 32):
+            return BinaryOp.int(
+                left=left_expr.left,
+                op="/",
+                right=Literal(new_denom),
+            )
+
+    # Detect `%`: (x - ((x / N) * N)) --> x % N
+    if expr.op == "-" and isinstance(right_expr, BinaryOp) and right_expr.op == "*":
+        div_expr = early_unwrap_ints(right_expr.left)
+        mod_base = early_unwrap_ints(right_expr.right)
+        if (
+            isinstance(div_expr, BinaryOp)
+            and early_unwrap_ints(div_expr.left) == left_expr
+            and div_expr.op == "/"
+            and early_unwrap_ints(div_expr.right) == mod_base
+            and isinstance(mod_base, Literal)
+        ):
+            return BinaryOp.int(left=left_expr, op="%", right=right_expr.right)
+
+    # Remove outer error term: ((x / N) - (x >> 31)) --> x / N
+    if (
+        expr.op == "-"
+        and isinstance(left_expr, BinaryOp)
+        and isinstance(right_expr, BinaryOp)
+        and right_expr.op == ">>"
+        and early_unwrap_ints(right_expr.right) == Literal(31)
+    ):
+        div_expr = left_expr
+        shift_var_expr = early_unwrap_ints(right_expr.left)
+        if isinstance(div_expr, BinaryOp) and div_expr.op == "/":
+            div_var_expr = early_unwrap_ints(div_expr.left)
+            # Check if the LHS of the shift is the same var that we're dividing by
+            if div_var_expr == shift_var_expr:
+                return div_expr
+            # If the var is under 32 bits, the error term may look like `(x << K) >> 31` instead
+            if (
+                isinstance(shift_var_expr, BinaryOp)
+                and early_unwrap_ints(div_expr.left)
+                == early_unwrap_ints(shift_var_expr.left)
+                and shift_var_expr.op == "<<"
+                and isinstance(shift_var_expr.right, Literal)
+            ):
+                return div_expr
+
+    # Shift on the result of the mul: (x >> M) / N --> x / (N << M)
+    if (
+        isinstance(left_expr, BinaryOp)
+        and expr.op == ">>"
+        and isinstance(right_expr, Literal)
+    ):
+        divisor_shift += right_expr.value
+        expr = left_expr
+        left_expr = early_unwrap_ints(expr.left)
+        right_expr = early_unwrap_ints(expr.right)
+
+    # Remove inner error term: MULT_HI(x, N) + x --> MULT_HI(x, N)
```

Comment:
This looks really weird to me. It MULT_HI(x, N) + x == MULT_HI(x, N + 2^32), I would have thought...

Reply:
I added some more comments for this step. Because `MULT_HI` is signed, if `N`'s highest bit is set, then it represents `(s32) ((u32) N - 2^32)` -- so adding back the `2^32` is sort of like calculating `(s32) x * (u32) N`.